compiler/src/dotty/tools/dotc/transform/UnusedArgLift.scala

```diff
@@ -2,17 +2,17 @@ package dotty.tools.dotc.transform
 
 import dotty.tools.dotc.ast.tpd
 import dotty.tools.dotc.core.Contexts._
-import dotty.tools.dotc.core.NameKinds._
+import dotty.tools.dotc.core.Symbols._
 import dotty.tools.dotc.core.Types._
 import dotty.tools.dotc.transform.MegaPhase.MiniPhase
 import dotty.tools.dotc.typer.EtaExpansion
 
 import scala.collection.mutable.ListBuffer
 
```

Comment:
We need to have a discussion whether we want to do this. The whole point of `unused` is to avoid generating code. So if we lift most arguments we miss that point. If we assume a precise effect analysis we could lift arguments if the argument is pure only. But we do not have such an effect analysis.

An alternative is to always drop unused arguments. In a sense unused arguments are then treated similarly to pattern guards - the compiler is allowed to assume they are pure. In the future, if we have a good effect analysis we could also _require_ that unused arguments are pure (and total?).

Reply:
Now I'm convinced that we should not evaluate any of the arguments in an unused method. This implies that unused implicits only check for the existence of such implicit and then disregard its computation. For unused arguments that are explicitly stated we should drop them as well but if they are not pure we will emit a warning.
We should not require them to be pure or total, but rather use those as hints in the warnings to point that proof might be unsafe.

```scala
def f(unused x: A): B = ???
def g()(unused implicit x: A): B = ???

f(pureA) // pureA is not evaluated
g() // the implicit A for x is not evaluated
g()(pureA) // pureA is not evaluated

f(impureA) // warn that impureA will not be evaluated
g()(impureA) // the implicit A for x is not evaluated
```